Adds importer 'privacy' param

gfiorav opened the pull request:
This will allow users with private tables enabled to set uploaded table's privacy setting to whatever thy specify in the `privacy` parameter of their request to the Importer API.

If the user submits an invalid privacy setting, an error is returned.
If the user doesn't have private tables enabled, an explanatory message is shown with a link to our [pricing plan](https://cartodb.com/pricing) is included.

This PR fixes #4945

saleiva commented:
This will need documentation, so please be sure @csobier is informed :)

gfiorav commented:
As soon as I fix a few things and revise I will write the docs. In the mean time @csobier, any suggestions on the format of the "You need to upgrade" message when a user can't have private tables?

csobier commented:
@gfiorav, this is hard to infer without seeing the action. Tag me when the dialogs are created and I will be happy to wordsmith!  (Here is my best guess but I know I am probably missing some valuable context): 

Invalid privacy setting dialog:
"The uploaded table privacy setting does not match the defined `privacy` parameter of your Import API request"

Disabled private tables dialog:
"You cannot upload the privacy setting as you do not have privacy tables enabled for your account. Contact [Sales](mailto:sales@cartodb.com) for more information"

Kartones commented:
AFAIK the only doc you should write is add the proper param here http://docs.cartodb.com/cartodb-platform/import-api.html#params no?

@danicarrion is going to love this ;)

Kartones commented:
I also see a small issue with this... table privacy is one thing, but there is also a parameter `create_vis` to automatically create a Map (visualization in backend terms) from the imported Dataset. Maybe would be nice to use this ticket to also define a `created_vis_privacy` parameter to allow fine control of the visualization's privacy. 

But can also go at another different ticket, just letting you all know that scenario exists

gfiorav commented:
@Kartones right now the privacy of the viz is the same as whatever the privacy param was set to. I can also add vis_privacy to control it explicitly! What do you think, @javisantana ? Also, thanks for the links :)

@csobier Thanks for the suggestions! :D These messages will be returned through the terminal, so It might be good to keep them short. Also, links have to be explicitly written out. Right now, the message for Invalid Privacy is more techie, it says: `Unknown value 'whatever_you_misspelt' for 'privacy'`.

javisantana commented:
about the vis_privacy: let's keep as simple as possible and have only a parameter

About the copy: please, API messages should be technical and provide useful information like:

"Unknown value '#{params[:privacy]}' for 'privacy'

can be

"Unknown value '#{params[:privacy]}' for 'privacy'. 'LINK', 'PRIVATE' and 'PUBLIC' are allowed"

gfiorav commented:
@javisantana done.

About the _you need to upgrade_ message... should this be technical too? The provisional message I used is `You can't create private tables yet! You need to upgrade. Check https://cartodb.com/pricing for more info.`

Sounds OK? Suggestions?

javisantana commented:
In my opinion API responses should not contain upgrade messages and should be really technical, like:

"your account type (FREE) does not allow to create private datasets.[Check https://cartodb.com/pricing for more info]"

and the api should return a 400 in that case.

javisantana commented:
@csobier can tell more about that probably

gfiorav commented:
Could you CR this for me @Kartones :)

Kartones commented on db/migrate/20150813151350_add_privacy_to_data_import.rb:
```diff
@@ -0,0 +1,13 @@
+Sequel.migration do
+  up do
+    alter_table :data_imports do
+    	add_column :privacy, Integer, :default => 0
```
we could have used this (right now fully unused) DataImport field: `import_extra_options`, which is a json... we're bloating a bit data_import for something that most users are not going to use (this is an advanced feature)...

But ok, just take into account for future fields, anything not going to be used frequently should go there.

Kartones replied:
Uhm but don't set default to `0`, leave as `NULL` by default. You're checking upward for a "if this is not null and user has private tables enabled blablabla", and there is Ruby logic to setup privacy currently, so don't change that flow with an optional param.

javisantana replied:
I think we can change this already, it's a matter of remove the migration and using the right field, right?

gfiorav replied:
Hi @juanignaciosl, can you leverage a bit on setting `default to 0`? I think you might be able to better explain why we thought we should set it that way.

Kartones replied:
Yep, if there is no hurry better to use the good approach, one less variable to remove in the future.

gfiorav replied:
Ok, I'll remove the migration and work with the better option of `import_extra_options`

Kartones commented on spec/connectors/importer_spec.rb:
```diff
@@ -6,10 +6,13 @@
 describe CartoDB::Connector::Importer do
 
-  before do
+  before(:each) do
```
mmm better approach:
in a `before(:all)` create the user, in a `after(:all)` destroy it, and use:
```ruby
after(:each) do delete_user_data(@user)
```

Creating and destroying users is expensive in test run time, so just cleaning the same user and reusing it helps save precious time

gfiorav replied:
Agreed.

Will change in upcoming push :+1:

Kartones commented on spec/connectors/importer_spec.rb:
```diff
@@ -77,10 +77,63 @@
+      :data_source   => filepath,
+      :updated_at    => Time.now,
+      :append        => false,
+      :privacy       => Carto::UserTable::PRIVACY_TEXTS_TO_VALUES['private']
```
Try also the scenario of not specifying privacy, with a user that has private tables enabled (should generate one private) and then with private tables disabled (should generate one public), so you make sure your changes are optional and don't change existing behaviour

gfiorav commented:
Retest this, please

gfiorav commented:
All comments by @Kartones and @javisantana have been taken into account in this new push. Please CR changes

ohasselblad commented:
I'm excited about this PR :)

juanignaciosl commented on app/controllers/api/json/imports_controller.rb:
```diff
@@ -131,7 +131,8 @@ def default_creation_options
       create_visualization:   ["true", true].include?(params[:create_vis]),
-      user_defined_limits:    user_defined_limits
+      user_defined_limits:    user_defined_limits,
+      import_extra_options:   import_extra_options
```
Don't reuse `import_extra_options`. This field comes from this commit, 5db66537b489b37065a5abd757802caaba86d314, and is only used as a stats storage (a shortcut to collect some analytical insights about the import). Indeed currently it only has one value, `{"common_data":true}`. We should document this somehow. @gfiorav if you see it fits somewhere around the code please do it. cc @Kartones .

juanignaciosl replied:
Oh, sorry, I didn't explain my point of view. This kind of catch-all storages are error-prone and really hard to mantain: what can it contain? Format? Order? Defaults? I think all around our schema there's one that make sense, `layers.options`, since it has no meaning for the backend, and it's just an storage for frontend options.

A `privacy` column is clear and straightforward.

Kartones replied:
@juanignaciosl  the problem with this is we're adding tons of params to data imports, and table privacy is an extra/optional param, which is only going to be used by advanced users. @gfiorav actually had setup a different param, but being optional and having this `import_extra_options` from my point of view is a nice fit for all non-mandatory ones.

You're right that it can cause more trouble, but so would do having a privacy field with non-null default value as the branch initially had, or a non-documented misleading field as it seems to happen with this extra options one.

I'm ok with both approaches, so ok to go back to a field, but only if:
- This extra options field has a proper documentation in the DataImport models and a more suitable rename (like `stats_options` or something similar)
- New field is `table_privacy` to avoid more issues (even if now we replicate privacy to the visualization, tomorrow we might not) and the default value is `NULL` to avoid breaking current behaviour.


We should start thinking on properly docummenting every model field (I don't care if with a JetBrains-like `# @param name type description` format or anything else) to avoid this things...

juanignaciosl replied:
I strongly agree with you on both requirements.

Not sure about field documentation, but it might be worth a try.

gfiorav replied:
Ok, so I will just revert to the original fix. Maybe @javisantana should also have a say so we can agree on something instead of going back and forth. I remember he had a say in the `import_extra_options` vs `privacy` dispute :smile:

juanignaciosl commented on app/controllers/api/json/imports_controller.rb:
```diff
@@ -179,4 +180,23 @@ def decrement_concurrent_imports_rate_limit
+    if params[:privacy].present?
+      privacy = UserTable::PRIVACY_TEXTS_TO_VALUES[params[:privacy].downcase]
+      raise "Unknown value '#{params[:privacy]}' for 'privacy'. 'private', 'public' and 'link' are allowed." if privacy.nil?
+      raise "Your account type (#{current_user.account_type.tr('[]','')}) does not allow to create private datasets. [Check https://cartodb.com/pricing for more info]" if privacy != UserTable::PRIVACY_PUBLIC && !current_user.private_tables_enabled
```
Controllers should not have business logic. Move this check to `User`: `current_user.valid_privacy?(privacy)`.

Also, is there any reason for using `[]` in the user message? I think "Check https://cartodb.com/pricing for more info." is better.

gfiorav replied:
Ok, I'll move that.

`[]` were suggested by @javisantana, but I guess it's not important, we can do whatever.

javisantana replied:
I didn't say anything about a [] haha

juanignaciosl commented on app/models/table.rb:
```diff
@@ -412,6 +412,13 @@ def after_create
 
+      if !@data_import.import_extra_options.nil?
+        extra_options = ::JSON.parse(@data_import.import_extra_options)
+        if !extra_options["privacy"].nil? && self.owner.private_tables_enabled
```
This check should also use `User` method after suggested refactor.

gfiorav replied:
:+1:

juanignaciosl commented on app/models/table.rb:
```diff
@@ -571,7 +578,7 @@ def create_derived_visualization
       }
     )
-    CartoDB::Visualization::Overlays.new(vis).create_default_overlays
+    CartoDB::Visualization::Overlays.new(vis).create_default_overlays  
```
Don´'t add trailing spaces.

gfiorav replied:
My bad

juanignaciosl commented on app/models/table/user_table.rb:
```diff
@@ -55,9 +55,15 @@ class UserTable < Sequel::Model
+    PRIVACY_LINK => 'link'
+  }
+
+  PRIVACY_TEXTS_TO_VALUES = {
```
Don't declare this twice. Declare the common one and use [Hash#invert](http://ruby-doc.org/core-1.9.3/Hash.html#method-i-invert).

Kartones replied:
Didn't knew this one, nice!

gfiorav replied:
Nice trick, didn't know

juanignaciosl commented on spec/connectors/importer_spec.rb:
```diff
@@ -77,10 +81,117 @@
+  end
+
+  it 'should import tables as public if privacy param is set to public' do
+    @user.private_tables_enabled = false
```
There're two other interesting tests you should do: import as pubic with `private_tables_enabled = true`, and check that request fails if `privacy` is set to `private` and `private_tables_enabled` is false.

Also, don't duplicate that much code:
1. Instead of creating a random file, use an existing fixture one at `spec/support/data` (or add a new simple one). Random filepath is not useful here and content is not dynamic, an existing file is simpler.
2. Use an existing factory (or create a new one) and use `FactoryGirl.create` instead of creating it by hand.

danicarrion commented:
sorry guys, just got back from vacation. i can't wait to see this in production!!!

gfiorav commented:
@csobier the provisional documentation for this feature can be found [here (PR #304 in /docs)](https://github.com/CartoDB/docs/pull/304). Would you be so kind to review it?

gfiorav commented:
See if everything is to your liking now, @juanignaciosl

juanignaciosl commented on app/controllers/api/json/imports_controller.rb:
```diff
@@ -131,7 +131,9 @@ def default_creation_options
       create_visualization:   ["true", true].include?(params[:create_vis]),
-      user_defined_limits:    user_defined_limits
+      user_defined_limits:    user_defined_limits,
+      import_extra_options:   import_extra_options,
```
Strange, why have you needed to add `import_extra_options`?

gfiorav replied:
`import_extra_options` was said to be used for adding any extra import options in the future, but was completely unused. Since it was suggested to use this for privacy, I developed it and then reverted back to using `privacy` I though that I could leave the infrastructure setup for the future.

If this is to be used, the only thing necessary is to handle extra params in JSON here: https://github.com/CartoDB/cartodb/blob/4945-importer-api-table-privacy/app/controllers/api/json/imports_controller.rb#L184-L191
and here: https://github.com/CartoDB/cartodb/blob/4945-importer-api-table-privacy/app/models/table.rb#L422-L424

It has also been tested, so why not just leave it there?

juanignaciosl replied:
I hadn't said you should remove it yet, just asking _why_ ;-) I have mixed feelings about `import_extra_options`, but this is not the place to discuss it, so I've opened #5082. I don't use to add changes I don't actually use, but if you've tested it and you're sure it works and it's better like this keep it. Moving `::JSON.dump(` from the model to the controller certainly smells fine.

juanignaciosl commented on app/controllers/api/json/imports_controller.rb:
```diff
@@ -179,4 +181,23 @@ def decrement_concurrent_imports_rate_limit
+  def privacy
+    if params[:privacy].present?
+      privacy = (UserTable::PRIVACY_VALUES_TO_TEXTS.invert)[params[:privacy].downcase]
+      raise "Unknown value '#{params[:privacy]}' for 'privacy'. 'private', 'public' and 'link' are allowed." if privacy.nil?
```
You're maybe going to kill me, but you shouldn't hardcode valid values. Change might not be worth, up to you.

gfiorav replied:
I originally made this dynamic but though it was a bit of an overkill :laughing: 

The code was:
```ruby
"Unknown value '#{params[:privacy]}' for 'privacy'. #{UserTable::PRIVACY_VALUES_TO_TEXTS.invert.keys.join(', ')} are allowed." if privacy.nil?
```

If you think it's better I'll add it :smile:

juanignaciosl replied:
I really dislike this kind of duplication and coupling, but it's up to you :) It doesn't feel like an overkill to me anyway.

juanignaciosl commented:
My only important concern is the need for `import_extra_options` related changes, the rest looks fine :+1:

gfiorav commented:
I reverted changes to `import_extra_options` as it isn't in the scope of this feature and will delay it otherwise. This will be taken care of in #5082

gfiorav commented:
@juanignaciosl 

I've made some changes:
* I reverted all `import_extra_options` modifications. The will be dealt with in #5082 
* I've added some minor things suggested

I've tested the functionality of the new feature in staging with every case I could imagine. I think it's ready.  Could you please give this a final thumbs up?

Once I have your final approval I will deploy this tomorrow morning.

Thanks :smiley: !

juanignaciosl commented:
:+1:

cesarandreslopez commented:
@gfiorav "@Kartones right now the privacy of the viz is the same as whatever the privacy param was set to. I can also add vis_privacy to control it explicitly! What do you think, @javisantana ? Also, thanks for the links :)"

Actually, i'm not seeing that behavoir in production. When you import through the import API and set the table as "public" and create_vis set to true the table will become public but the vizualization will remain private.
